rpcmetrics/endpoints.go

```diff
@@ -0,0 +1,77 @@
+package rpcmetrics
+
+import "sync"
+
+type normalizedEndpoints struct {
+	names       map[string]string
+	maxSize     int
+	defaultName string
+	mux         sync.RWMutex
+}
+
+func newNormalizedEndpoints(maxSize int, defaultName string) *normalizedEndpoints {
+	return &normalizedEndpoints{
+		maxSize:     maxSize,
+		defaultName: defaultName,
+		names:       make(map[string]string, maxSize+1),
+	}
+}
+
+func (n *normalizedEndpoints) normalize(name string) string {
+	n.mux.RLock()
+	norm := n.names[name]
+	n.mux.RUnlock()
+	if norm != "" {
+		return norm
+	}
+	norm = n.safeName(name)
+	n.mux.Lock()
+	defer n.mux.Unlock()
+	if norm, ok := n.names[name]; ok {
+		// already added after we released read lock
+		return norm
+	}
+	if len(n.names) >= n.maxSize {
+		return n.defaultName
+	}
+	n.names[name] = norm
+	return norm
+}
+
+// safeName ensures that the name is in the safe charset [a-zA-Z0-9-./_].
+// Any characters not in this set are replaced with '-'.
+func (n *normalizedEndpoints) safeName(name string) string {
```

Comment:
in theory the allowed charset could be provided externally, in which case these will need to be proper methods